File: frontend/src/reducers/property.js
const initialState = {
  error: "",
  addProperty: {
    fields: {
      landlord_id: "",
      navigator_id: "",
      property_name: "",
      address: "",
      unit_apt_no: "",
      floor: 0,
      zip_code: "",
      housing_type: "",
      bedrooms: 0,
      bathrooms: 0,
      shared_bathrooms: 0,
      has_basement: false,
      school_district: "",
      bus_line: false,
      wheelchair_accessibility: false,
      elevator: false,
      date_first_available: new Date().toISOString().slice(0, 10),
      year_available: new Date().getFullYear(),
      potential_month_available: 0,
      listing_date: new Date().toISOString().slice(0, 10),
      where_listed: "",
      contact_method: [],
      last_contact_date: new Date().toISOString().slice(0, 10),
      last_contacted_by: "",
      monthly_rent: 0,
      application_fee: 0,
      deposit: 0,
      last_month_rent_required: false,
      voucher_type_accepted: [],
      voucher_type_not_accepted: [],
      credit_screening_company: "",
      background_screening_company: "",
      allow_criminal_records: false,
      is_available: false,
      notes: "",
    },
  },
};

const propertyState = (state = initialState, action) => {
  switch (action.type) {
    case "SET_NEWPROP_SUCCESS":
      return Object.assign({}, state, {
        ...state,
        id: action.id,
        error: "",
      });
    case "SET_NEWPROP_FAILURE":
      return Object.assign({}, state, {
        ...state,
        error: action.error,
      });
    case "SET_ADD_PROPERTY_FIELD":
      if (Array.isArray(state.addProperty.fields[action.fieldName])) {
        let index = state.addProperty.fields[action.fieldName].indexOf(
          action.newValue
        );
        if (index > -1) {
          return Object.assign({}, state, {
            ...state,
            addProperty: {
              fields: {
                ...state.addProperty.fields,
                [action.fieldName]: state.addProperty.fields[
                  action.fieldName
                ].filter((item) => item !== action.newValue),
              },
            },
          });
        } else {
          return Object.assign({}, state, {
            ...state,
            addProperty: {
              fields: {
                ...state.addProperty.fields,
                [action.fieldName]: [
                  ...state.addProperty.fields[action.fieldName],
                  action.newValue,
                ],
              },
            },
          });
        }
      } else {
        return Object.assign({}, state, {
          ...state,
          addProperty: {
            fields: {
              ...state.addProperty.fields,
              [action.fieldName]: action.newValue,
            },
          },
        });
      }
    default:
      return state;
  }
};

export default propertyState;

File: backend/src/models/property.py
import json

from marshmallow_sqlalchemy import SQLAlchemyAutoSchema
from sqlalchemy import Column
from sqlalchemy.dialects.postgresql import (ARRAY, BOOLEAN, DATE, FLOAT,
                                            INTEGER, UUID, VARCHAR)

from models.base import Base


class Property(Base):
  __tablename__ = 'property'

  id = Column(UUID, primary_key = True)
  landlord_id = Column(UUID)
  navigator_id = Column(UUID)
  voucher_type_accepted = Column(ARRAY(VARCHAR))
  voucher_type_not_accepted = Column(ARRAY(VARCHAR))
  address = Column(VARCHAR)
  zip_code = Column(VARCHAR)
  unit_apt_no = Column(VARCHAR)
  property_name = Column(VARCHAR)
  bus_line = Column(BOOLEAN)
  school_district = Column(VARCHAR)
  wheelchair_accessibility = Column(BOOLEAN)
  elevator = Column(BOOLEAN)
  monthly_rent = Column(INTEGER)
  contact_method = Column(ARRAY(VARCHAR))
  is_available = Column(BOOLEAN)
  date_first_available = Column(DATE)
  last_contact_date = Column(DATE)
  potential_month_available = Column(INTEGER)
  bedrooms = Column(INTEGER)
  bathrooms = Column(INTEGER)
  shared_bathrooms = Column(INTEGER)
  has_basement = Column(BOOLEAN)
  application_fee = Column(FLOAT)
  deposit = Column(FLOAT)
  last_month_rent_required = Column(BOOLEAN)
  allow_criminal_records = Column(BOOLEAN)
  listing_date = Column(DATE)
  where_listed = Column(ARRAY(VARCHAR))
  floor = Column(INTEGER)
  housing_type = Column(VARCHAR)
  year_available = Column(INTEGER)
  credit_screening_company = Column(VARCHAR)
  background_screening_company = Column(VARCHAR)
  last_contacted_by = Column(UUID)
  pets_allowed = Column(BOOLEAN)
  background_check_required = Column(BOOLEAN)
  near_busstop = Column(BOOLEAN)
  primary_contact = Column(VARCHAR)
  contact_phone_number = Column(VARCHAR)

  def __repr__(self):
    _dict = PropertySchema().dump(self)
    return json.dumps(_dict)

class PropertySchema(SQLAlchemyAutoSchema):
  class Meta:
    model = Property
